jkube-kit/config/service/src/main/java/org/eclipse/jkube/kit/config/service/DebugService.java

```diff
@@ -0,0 +1,318 @@
+    Map<String, String> initDebugEnvVarsMap(boolean debugSuspend) {
+        Map<String, String> envVars = new TreeMap<>();
+        envVars.put(DebugConstants.ENV_VAR_JAVA_DEBUG, "true");
+        envVars.put(DebugConstants.ENV_VAR_JAVA_DEBUG_SUSPEND, String.valueOf(debugSuspend));
+        if (this.debugSuspendValue != null) {
+            envVars.put(DebugConstants.ENV_VAR_JAVA_DEBUG_SESSION, this.debugSuspendValue);
+        }
+        return envVars;
+    }
+
+    // Visible for testing
+    void enableDebugging(HasMetadata entity, String fileName, boolean debugSuspend) {
+      if (entity instanceof Deployment) {
+        enableDebugging((Deployment) entity, fileName, debugSuspend);
+      } else if (entity instanceof ReplicaSet) {
+        enableDebugging((ReplicaSet) entity, fileName, debugSuspend);
+      } else if (entity instanceof ReplicationController) {
+        enableDebugging((ReplicationController) entity, fileName, debugSuspend);
+      } else if (entity instanceof DeploymentConfig) {
+        enableDebugging((DeploymentConfig) entity, fileName, debugSuspend);
+      }
+    }
+
+    private Consumer<PodTemplateSpec> applyEntity(HasMetadata entity, String fileName) {
+      return template -> {
+        log.info(DEBUG_ENV_VARS_UPDATE_MESSAGE, entity.getKind(), entity.getMetadata().getName());
+        applyService.apply(entity, fileName);
+      };
+    }
+
+    private Predicate<PodTemplateSpec> enableDebuggingFilterFunc(HasMetadata entity, boolean debugSuspend) {
+        return pts -> enableDebugging(entity, pts, debugSuspend);
+    }
+
+    private boolean isDebugAlreadyEnabled(ReplicationController entity, boolean debugSuspend) {
+      return firstContainerHasEnvVars(
+          kubernetesClient.resource(entity).fromServer().get().getSpec().getTemplate().getSpec().getContainers(),
+          initDebugEnvVarsMap(debugSuspend));
+    }
+
+    private boolean isDebugAlreadyEnabled(DeploymentConfig entity, boolean debugSuspend) {
+        return firstContainerHasEnvVars(
+            kubernetesClient.resource(entity).fromServer().get().getSpec().getTemplate().getSpec().getContainers(),
+            initDebugEnvVarsMap(debugSuspend));
+    }
+
+    private void enableDebugging(ReplicationController entity, String fileName, boolean debugSuspend) {
+      Optional.ofNullable(entity)
+          .map(ReplicationController::getSpec)
+          .map(ReplicationControllerSpec::getTemplate)
+          .filter(enableDebuggingFilterFunc(entity, debugSuspend))
+          .filter(pts -> !isDebugAlreadyEnabled(entity, debugSuspend))
+          .ifPresent(applyEntity(entity, fileName));
+    }
+
+    private void enableDebugging(ReplicaSet entity, String fileName, boolean debugSuspend) {
+      Optional.ofNullable(entity)
+          .map(ReplicaSet::getSpec)
+          .map(ReplicaSetSpec::getTemplate)
+          .filter(enableDebuggingFilterFunc(entity, debugSuspend))
+          .ifPresent(applyEntity(entity, fileName));
```

Comment:
I tested but the debug goal doesn't seem to be working for me. Looks like environment variables are not added to pods on updating ReplicSet automatically, they only get added when I manually delete the old pod.

I think `ReplicaSet` resource doesn't support rolling-update. As per Kubernetes [docs](https://kubernetes.io/docs/concepts/workloads/controllers/replicaset/#deleting-just-a-replicaset):

> As long as the old and new .spec.selector are the same, then the new one will adopt the old Pods. However, it will not make any effort to make existing Pods match a new, different pod template. To update Pods to a new spec in a controlled way, use a Deployment, as ReplicaSets do not support a rolling update directly.

Reply:
See my comment in the issue description.
No, ReplicaSets don't support rolling updates (or any other workaround). They require existent pods to be deleted when applied. The delete-pods functionality (which on the other hand **is** provided for ReplicationControllers) is not implemented in ApplyService for ReplicaSets (this is obviously out of scope for this issue, and probably an unneeded feature).
![image](https://user-images.githubusercontent.com/488391/104355015-5809c000-550a-11eb-96e4-e7d9ba530fe3.png)